File: leetcode/go/330.patching-array.go
func minPatches(nums []int, n int) int {
    missing, ans, i, l := 1, 0, 0, len(nums)
    for missing <= n {
        if i < l && nums[i] <= missing {
            missing += nums[i]
            i++
        } else {
            missing <<= 1
            ans++
        }
    }
    return ans
}

File: leetcode/rust-leetcode/src/330.patching-array.rs
impl Solution {
    pub fn min_patches(nums: Vec<i32>, n: i32) -> i32 {
        let n = n as i64;
        let nums = nums.iter().map(|&n| n as i64).collect::<Vec<_>>();
        let (mut missing, mut i, mut ans) = (1, 0, 0);
        while missing <= n {
            if i < nums.len() && nums[i] <= missing {
                missing += nums[i];
                i += 1;
            } else {
                missing <<= 1;
                ans += 1;
            }
        }
        ans
    }
}

